### ISSUE TITLE
A privacy friendly referral program.

### ISSUE THREAD

**Issue by Reckless-Satoshi** [RoboSats/robosats#38]
Eventually the usefulness of RoboSats will boil down to how much liquidity is in the book. 

In order to increase liquidity organically (i.e., make the platform more popular) one could think of a referral program; that is, a user is rewarded for recommending the platform to other potential users. However, implementing a referral program is extremely difficult while 1) avoiding users to spy on each other and 2) keeping minimal tracking and logs.

**Example Implementations**

1. Reward a user X% from the fee for every trade of the referred users (aka Binance model). **Issue**: can easily be exploited by a user to spy on other users' trading activity, especially if the rewards are deterministically linked to the users activity.

2. Direct single payout for every new user that is referred. **Issue**: the low barrier to create new users in RoboSats makes it very easy to abuse this feature. One can create a referral link and reward himself many times.

I do not think there is an optimal solution, therefore a referral program will probably never be implemented. However, it makes for an interesting thinking puzzle... Maybe there is a way to solve it...

**Comment by Reckless-Satoshi** [comment]
**It is clear a method for organic growth is needed**. Referrals are the most appropriated ones. I will be working on implementing a referral program starting today.

- [ ] Profile pop up shows: your robot's referral link and rewards accumulated
- [ ] Tapping on rewards allow you to submit a LN invoice to withdraw them.
- [ ] A new user landing in the homepage from a referral link will create a robot that `is_referred=true`
- [ ] At trade time, the new Robot has an extra flat fee (e.g. 100 Sats)
- [ ] Everyday midnight rewards are added up to the balance of the robots who referred.

Why this set up? Mostly to limit spying with referral links and avoid exploits:

1. **Flat reward:** so those who send link cannot know the size of the trades.

2. **Balance reflected at midnight:** so those who send link cannot know when the trades happened

3. **The new user tips his referee:** so this feature cannot be exploited to make the platform run into loses by automating bot creation and self-referrals.

Of course, this is just a start. Most important is to implement the system. Parameters can be changed later :)

**Comment by ibertario** [comment]
Some improvements that can make the system stronger:

* Reward ID is created only after a trade is completed
* You obtain a CODE, a shareable LINK (..../ref/CODE) and an ADMIN REFERRAL LINK
* With the admin referral link, you can see what you have obtained, withdraw... and maybe donate the sats to RoboSats.

This solves the issue: [Referral codes persistence and links with Robots](https://github.com/Reckless-Satoshi/robosats/issues/91#) and prevents some marginal Robots reuse.

A little T&C can be added (about the amount of time the earnings are saved without any admin referral link login, security implications, etc.)

**Comment by Reckless-Satoshi** [comment]
Yay! Good ideas, thanks for the brainstorm.

> * Reward ID is created only after a trade is completed

It's preferable to not have to complete a trade to obtain a link. It is not trivial to complete a trade and we do not want to incentivice self-trading to create referral links. In addition, it opens the door to potentially be able to link a real ID with a referral code... Not great for privacy. Better if a robot can be used for referrals only.

> * You obtain a CODE, a shareable LINK (..../ref/CODE) and an ADMIN REFERRAL LINK

An admin referral link implies an admin panel and a UI element. Too much work as of now. Probably on a future version, not a priority as of now.

> * With the admin referral link, you can see what you have obtained, withdraw... and maybe donate the sats to RoboSats.

Pretty cool!

**Comment by ibertario** [comment]
Very good feedback, thank you!

> Better if a robot can be used for referrals only.

This can be a good way to do the work!

**Comment by Reckless-Satoshi** [comment]
At the moment the referral program has been totally dismantled as it is incompatible with the upcoming federated frontend (many coordinators, who will tip who and why?)

**Comment by kiz8** [comment]
it's a shame this was closed, but it's quite a difficult problem. Pulling it off would provide a mechanism to exponentially grow adoption, so let me share a few brainwaves:

- Increase legitimate usage of referral system: the 100 sat tip from referal to referer wasn't motivating anyone to promote anything. take a small portion of coordinator/robosats fees from the first trade to create a juicier referral reward. 

- Federated coordinators could opt-in to receiving referred users (in exchange for contributing to the referral fees) for their first trade. Coordinators are then betting on "first contact stickiness" to keep that new user as a long time client of their coordinator. Increases competition among coordinators.

- 30-day wait on referral reward withdrawals: to have a window for catching clearly fraudulent behavior, sybil or otherwise. (unclear how to exactly detect this, legitimate referrers would likely be okay with some vetting though). 

- Special incentives: to foster growth in highly desired markets (say LN swaps or UK p2p or whatever sector earns best for Robosats/coordinators)
